### PR TITLE
Add CBMC proof for IPv6 TCPReturnPacket 

### PR DESCRIPTION
**Pull request by moninom1**
Description
-----------
Add changes to validate CBMC proof for IPv6 TCPReturnPacket

Test Steps
-----------
CBMC proof :
```
python prepare.py
make
```

Checklist:
----------

- [x] I have tested my changes. No regression in existing tests.
- [x] I have modified and/or added unit-tests to cover the code changes in this Pull Request.

Related Issue
-----------

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: test/cbmc/proofs/TCP/prvTCPReturnPacket_IPv6/Makefile.json [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+# NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+# BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+# ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+# CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+#
+# http://aws.amazon.com/freertos
+# http://www.FreeRTOS.org
+#
+
```
Comment (ActoryOu) [inline test/cbmc/proofs/TCP/prvTCPReturnPacket_IPv6/Makefile.json:28]:
```suggestion
```
We don't need these comments in json file.

File: test/cbmc/proofs/TCP/prvTCPReturnPacket_IPv6/TCPReturnPacket_IPv6_harness.c [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+/* FreeRTOS includes. */
+#include "FreeRTOS.h"
+#include "queue.h"
+
+/* FreeRTOS+TCP includes. */
+#include "FreeRTOS_IP.h"
+#include "FreeRTOS_IP_Private.h"
+#include "FreeRTOS_TCP_IP.h"
+#include "FreeRTOS_TCP_Transmission.h"
+
```
Comment (ActoryOu) [inline test/cbmc/proofs/TCP/prvTCPReturnPacket_IPv6/TCPReturnPacket_IPv6_harness.c:38]:
```suggestion

/* CBMC includes. */
```

File: test/cbmc/proofs/TCP/prvTCPReturnPacket_IPv6/TCPReturnPacket_IPv6_harness.c [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+        if( ensure_memory_is_valid( pxNetworkBuffer->pxEndPoint, sizeof( NetworkEndPoint_t ) ) )
+        {
+            /* Add an interface */
+            pxNetworkBuffer->pxEndPoint->pxNetworkInterface = ( NetworkInterface_t * ) malloc( sizeof( NetworkInterface_t ) );
+            __CPROVER_assume( pxNetworkBuffer->pxEndPoint->pxNetworkInterface != NULL );
+            pxNetworkBuffer->pxEndPoint->pxNetworkInterface->pfOutput = NetworkInterfaceOutputFunction_Stub;
+        }
+    }
+    /* If not. */
+    else
```
Comment (ActoryOu) [inline test/cbmc/proofs/TCP/prvTCPReturnPacket_IPv6/TCPReturnPacket_IPv6_harness.c:169]: Is it possible to enter this `else` condition?

File: test/cbmc/proofs/TCP/prvTCPReturnPacket_IPv6/TCPReturnPacket_IPv6_harness.c [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+    /* The code does not expect both of these to be equal to NULL at the same time. */
+    __CPROVER_assume( pxSocket != NULL || pxNetworkBuffer != NULL );
+
+    /* If network buffer is properly created. */
+    if( ensure_memory_is_valid( pxNetworkBuffer, sizeof( *pxNetworkBuffer ) ) )
+    {
+        /* Assume that the length is proper. */
+        __CPROVER_assume( ( ulLen >= sizeof( TCPPacket_IPv6_t ) ) && ( ulLen < ipconfigNETWORK_MTU ) );
+        pxNetworkBuffer->pucEthernetBuffer = safeMalloc( ulLen + ipSIZE_OF_ETH_HEADER );
+        __CPROVER_assume( pxNetworkBuffer->pucEthernetBuffer );
```
Comment (ActoryOu) [inline test/cbmc/proofs/TCP/prvTCPReturnPacket_IPv6/TCPReturnPacket_IPv6_harness.c:152]:
```suggestion
        __CPROVER_assume( pxNetworkBuffer->pucEthernetBuffer != NULL );
```

File: test/cbmc/proofs/TCP/prvTCPReturnPacket_IPv6/TCPReturnPacket_IPv6_harness.c [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+
+        pxNetworkBuffer->xDataLength = ( size_t ) ulLen;
```
Comment (ActoryOu) [inline test/cbmc/proofs/TCP/prvTCPReturnPacket_IPv6/TCPReturnPacket_IPv6_harness.c:154]:
```suggestion
        pxNetworkBuffer->xDataLength = ( size_t )( ulLen + ipSIZE_OF_ETH_HEADER );
```
Why are data length and allocate size different?

**Review by ActoryOu — CHANGES_REQUESTED** [review]

File: test/cbmc/proofs/TCP/prvTCPReturnPacket_IPv6/TCPReturnPacket_IPv6_harness.c (hunk 3 above)
Comment (moninom1) [reply test/cbmc/proofs/TCP/prvTCPReturnPacket_IPv6/TCPReturnPacket_IPv6_harness.c:170]: Yes ensure_FreeRTOS_NetworkBuffer_is_allocated uses safeMalloc

Comment (ActoryOu) [reply test/cbmc/proofs/TCP/prvTCPReturnPacket_IPv6/TCPReturnPacket_IPv6_harness.c:170]: ensure_memory_is_valid is used to check the network buffer descriptor only.
And it's assumed to be non-null in line 144. Is it still necessary to check it again?

```
#define ensure_memory_is_valid( px, length )    ( px != NULL ) && __CPROVER_w_ok( ( px ), length )
```

line 144:
```
__CPROVER_assume( pxSocket != NULL || pxNetworkBuffer != NULL );
```

Comment (moninom1) [reply test/cbmc/proofs/TCP/prvTCPReturnPacket_IPv6/TCPReturnPacket_IPv6_harness.c:170]: It is assumed both can not be null at the same time, but there is still possibility of pxNetworkBuffer to be NULL when pxSocket is not NULL.

**Review by ActoryOu — DISMISSED** [review]

**Pull request merged**